pkg/scaler/prober.go

```diff
@@ -0,0 +1,324 @@
+		secret, err = p.secretInterface.Get(secretName, metav1.GetOptions{})
+		if apierrors.IsNotFound(err) {
+			return nil, err
+		}
+		if err == nil {
+			break
+		}
+	}
+
+	if err != nil {
+		return nil, err
+	}
+
+	kubeconfig := secret.Data["kubeconfig"]
+	if kubeconfig == nil {
+		return nil, errors.New("Invalid empry kubeconfig")
+	}
+
+	clientConfig, err := clientcmd.NewClientConfigFromBytes(kubeconfig)
+	if err != nil {
+		return nil, err
+	}
+
+	config, err := clientConfig.ClientConfig()
+	if err != nil {
+		return nil, err
+	}
+
+	config.Timeout = toDuration(p.probeDeps.Probe.TimeoutSeconds, defaultTimeoutSeconds)
+
+	return kubernetes.NewForConfig(config)
+}
+
+// runWorker is a long-running function that will continually call the
+// processNextWorkItem function in order to read and process a message on the
+// workqueue.
+func (p *prober) run(stopCh <-chan struct{}) error {
+	if p == nil || p.probeDeps == nil || p.probeDeps.Probe == nil {
+		return errors.New("Invalid empty probe dependants configuration")
+	}
+	if p.probeDeps.Probe.External == nil {
+		return errors.New("Invalid empty external probe configuration")
+	}
+	if p.probeDeps.Probe.Internal == nil {
+		return errors.New("Invalid empty internal probe configuration")
+	}
+
+	p.initialDelay = toDuration(p.probeDeps.Probe.InitialDelaySeconds, defaultInitialDelaySeconds)
+
+	if p.probeDeps.Probe.SuccessThreshold != nil {
+		p.successThreshold = *p.probeDeps.Probe.SuccessThreshold
+	} else {
+		p.successThreshold = defaultSuccessThreshold
+	}
+
+	if p.probeDeps.Probe.FailureThreshold != nil {
+		p.failureThreshold = *p.probeDeps.Probe.FailureThreshold
+	} else {
+		p.failureThreshold = defaultFailureThreshold
+	}
+
+	var err error
+	p.internalClient, err = p.getClientFromSecret(p.probeDeps.Probe.Internal.KubeconfigSecretName)
+	if err != nil {
+		return err
+	}
+	p.externalClient, err = p.getClientFromSecret(p.probeDeps.Probe.External.KubeconfigSecretName)
+	if err != nil {
+		return err
+	}
+
+	if p.resultCh != nil {
+		close(p.resultCh)
+	}
+
+	p.resultCh = make(chan *probeResult)
+
+	ticker := time.NewTicker(toDuration(p.probeDeps.Probe.PeriodSeconds, defaultPeriodSeconds))
+	defer ticker.Stop()
+
+	for {
+		select {
+		case <-stopCh:
+			return nil
+		case <-ticker.C:
+			if p.initialDelayTimer != nil {
+				<-p.initialDelayTimer.C
+				p.initialDelayTimer.Stop()
+				p.initialDelayTimer = nil
+			}
+			if err := p.probe(); err != nil {
+				return err
+			}
+		}
+	}
+}
+
+func toDuration(seconds *int32, defaultSeconds int32) time.Duration {
+	if seconds != nil {
+		return time.Duration(*seconds) * time.Second
+	}
+	return time.Duration(defaultSeconds) * time.Second
+}
+
+func (p *prober) isHealthy(pr *probeResult) bool {
+	return pr.lastError == nil && pr.resultRun >= p.successThreshold
+}
+
+func (p *prober) isUnhealthy(pr *probeResult) bool {
+	return pr.lastError != nil && pr.resultRun >= p.failureThreshold
+}
+
+func (p *prober) probe() error {
```

Comment:
Comment on the implementation would be good as well

Reply:
Added comments to document the logic.